associated-token-account/program/src/processor.rs

```diff
@@ -154,3 +160,132 @@ fn process_create_associated_token_account(
+    let (owner_associated_token_address, bump_seed) =
+        get_associated_token_address_and_bump_seed_internal(
+            wallet_account_info.key,
+            owner_token_mint_info.key,
+            program_id,
+            spl_token_program_id,
+        );
+    if owner_associated_token_address != *owner_associated_token_account_info.key {
+        msg!("Error: Owner associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    // Check nested address derivation
+    let nested_associated_token_address = get_associated_token_address_with_program_id(
+        owner_associated_token_account_info.key,
+        nested_token_mint_info.key,
+        spl_token_program_id,
+    );
+    if nested_associated_token_address != *nested_associated_token_account_info.key {
+        msg!("Error: Nested associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    let owner_associated_token_account_signer_seeds: &[&[_]] = &[
+        &wallet_account_info.key.to_bytes(),
+        &spl_token_program_id.to_bytes(),
+        &owner_token_mint_info.key.to_bytes(),
+        &[bump_seed],
+    ];
+
+    if !wallet_account_info.is_signer {
+        msg!("Wallet of the owner associated token account must sign");
+        return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    if owner_token_mint_info.owner != spl_token_program_id {
+        msg!("Owner mint not owned by provided token program");
+        return Err(ProgramError::IllegalOwner);
+    }
+
+    // Account data is dropped at the end of this, so the CPI can succeed
+    // without a double-borrow
+    let (amount, decimals) = {
+        // Check owner associated token account data
+        if owner_associated_token_account_info.owner != spl_token_program_id {
+            msg!("Owner associated token account not owned by provided token program, recreate the owner associated token account first");
+            return Err(ProgramError::IllegalOwner);
+        }
+        let owner_account_data = owner_associated_token_account_info.data.borrow();
+        let owner_account = StateWithExtensions::<Account>::unpack(&owner_account_data)?;
+        if owner_account.base.owner != *wallet_account_info.key {
+            msg!("Owner associated token account not owned by provided wallet");
+            return Err(AssociatedTokenAccountError::InvalidOwner.into());
+        }
+
+        // Check nested associated token account data
+        if nested_associated_token_account_info.owner != spl_token_program_id {
+            msg!("Nested associated token account not owned by provided token program");
+            return Err(ProgramError::IllegalOwner);
+        }
+        let nested_account_data = nested_associated_token_account_info.data.borrow();
+        let nested_account = StateWithExtensions::<Account>::unpack(&nested_account_data)?;
+        let amount = nested_account.base.amount;
```

Comment:
why not checking `if nested_account.base.owner != *owner_associated_token_account_info.key`?

note that there could be arbitrary state at `nested_associated_token_account_info.key`. (ref: https://github.com/solana-labs/solana-program-library/pull/2889/files#r805334901)

Reply:
The idea is that `spl-token` will check that during the transfer, happy to add the check for total safety though.